src/w3d/renderer/scene.cpp

```diff
@@ -0,0 +1,264 @@
+            DX8Wrapper::Set_DX8_Render_State(D3DRS_FILLMODE, D3DFILL_WIREFRAME);
+            DX8Wrapper::Set_DX8_Render_State(D3DRS_ZBIAS, 7); // BFME2 calls Set_DX8_ZBias(7.0f) here
+#endif
+            Customized_Render(rinfo);
+        }
+
+        W3D::Enable_Texturing(state);
+    }
+
+    Post_Render_Processing(rinfo);
+}
+
+// 0x00832890
+SimpleSceneClass::SimpleSceneClass() : m_visibilityChecked(false) {}
+
+// 0x00832B70
+void SimpleSceneClass::Add_Render_Object(RenderObjClass *obj)
+{
+    SceneClass::Add_Render_Object(obj);
+    m_renderList.Add(obj);
+}
+
+// 0x00832BB0
+void SimpleSceneClass::Remove_Render_Object(RenderObjClass *obj)
+{
+    SceneClass::Remove_Render_Object(obj);
+    m_renderList.Remove(obj);
+}
+
+// 0x00833160
+SceneIterator *SimpleSceneClass::Create_Iterator(bool onlyvisible)
+{
+    // onlyvisible is not used!
+    return new SimpleSceneIterator(&m_renderList, onlyvisible);
+}
+
+void SimpleSceneClass::Destroy_Iterator(SceneIterator *it)
+{
+    delete it;
+}
+
+// 0x00832BF0
+void SimpleSceneClass::Register(RenderObjClass *obj, RegType for_what)
+{
+    switch (for_what) {
+        case RegType::ON_FRAME_UPDATE:
+            m_updateList.Add(obj);
+            break;
+        case RegType::LIGHT:
+            m_lightList.Add(obj);
+            break;
+        case RegType::RELEASE:
+            m_releaseList.Add(obj);
+            break;
+        default:
+            break;
+    }
+}
+
+// 0x00832C80
+void SimpleSceneClass::Unregister(RenderObjClass *obj, RegType for_what)
+{
+    switch (for_what) {
+        case RegType::ON_FRAME_UPDATE:
+            m_updateList.Remove(obj);
+            break;
+        case RegType::LIGHT:
+            m_lightList.Remove(obj);
+            break;
+        case RegType::RELEASE:
+            m_releaseList.Remove(obj);
+            break;
+        default:
+            break;
+    }
+}
+
+// 0x00832DF0
+float SimpleSceneClass::Compute_Point_Visibility(RenderInfoClass &rinfo, const Vector3 &point)
+{
+    CastResultStruct res;
+    auto camera_position = rinfo.m_camera.Get_Position();
+    LineSegClass ray{ camera_position, point };
+    RayCollisionTestClass raytest(ray, &res);
+
+    for (auto iter = m_renderList.Iterator(); iter; ++iter) {
+        auto *robj = iter.Peek_Obj();
+        robj->Cast_Ray(raytest);
+    }
+    return res.fraction == 1.0f ? 1.0f : 0.0f;
+}
+
+// 0x00832F40
+void SimpleSceneClass::Customized_Render(RenderInfoClass &rinfo)
+{
+    if (m_visibilityChecked == false) {
+        Visibility_Check(&rinfo.m_camera);
+    }
+
+    m_visibilityChecked = false;
+
+    for (auto iter = m_updateList.Iterator(); iter; ++iter) {
+        auto *robj = iter.Peek_Obj();
+        robj->On_Frame_Update();
+    }
+
+#ifdef BUILD_WITH_D3D8
+    for (auto i = 0; i < LightEnvironmentClass::MAX_LIGHTS; ++i) {
+        DX8Wrapper::Set_Light(i, nullptr);
+    }
+#endif
+
+    auto *light_environment = rinfo.m_lightEnvironment;
+    if (light_environment != nullptr) {
```

Comment:
should assert rinfo.light_environment==NULL before this and this if case should be  if ( !light_environment  ) as it makes a fallback light if one isn't set